src/hotspot/share/runtime/mutex.hpp

```diff
@@ -50,9 +50,9 @@ class Mutex : public CHeapObj<mtSynchronizer> {
        tty            = stackwatermark +   3,
        special        = tty            +   3,
        oopstorage     = special        +   3,
-       leaf           = oopstorage     +  10,
-       safepoint      = leaf           +  10,
-       barrier        = safepoint      +   1,
+       nosafepoint    = oopstorage     +   6,
```

Comment:
Maybe add a comment below `nosafepoint` like:

```
// A thread is not allowed to safepoint while holding a mutex whose
// rank is nosafepoint or lower.
```

Also, how about renaming the generic name `lock_types` to something specific like `standard_lock_ranks`?


BTW, should we (in separate RFE) add a new enum `_safepoint_check_default`, so that this parameter can be omitted depending on the rank value (unless in places where you need to override it)?

For one thing, I never understood which _safepoint_check_xxx I should have used when adding a new lock. I just randomly changed it until the JVM stops crashing.

Reply:
The comment is good, but I added it to above the enum because otherwise it messes up my nice alignment.

How about changing lock_types to Rank, since the later plan is to make this an enum class that only allows subtraction operations?

I think in a future RFE we should remove this safepoint_check_always or never parameter and just use the ranks, since after the next change there won't be any overrides.

For the most part, new locks that JavaThreads or runtime JRT entries use, should be safepoint_check_always so that the safepoint protocol is respected.  More locks are safepoint_check_never since these locks are shared with GC or compiler threads and many times used during a safepoint.